mypy/plugins/dataclasses.py

```diff
@@ -258,8 +261,15 @@ def collect_attributes(self) -> List[DataclassAttribute]:
 
                     known_attrs.add(name)
                     super_attrs.append(attr)
-
-        all_attrs = super_attrs + attrs
+                else:
+                    # How early in the attribute list an attribute appears is determined by the
+                    # reverse MRO, not simply MRO.
+                    # See https://docs.python.org/3/library/dataclasses.html#inheritance for
+                    # details.
+                    (attr,) = [a for a in all_attrs if a.name == name]
```

Comment:
I don't believe it can happen within a single class and with multiple classes the `if name not in known_attrs` block either adds an attribute that wasn't there or, in case of an attribute being overridden, it reorders attributes without adding duplicates.

The way this assignment is written is to express the fact that we strictly expect single attribute with the `name` name here, I can't think of a situation in which this condition is not met, any possible name clashes are covered by the tests already.